File: kotlin/advent/src/main/kotlin/org/gristle/pdxpuzzles/advent/y2020/Y20D14.kt
package org.gristle.pdxpuzzles.advent.y2020

import org.gristle.pdxpuzzles.advent.utilities.Day

class Y20D14(input: String) : Day {

    sealed class Instruction {

        companion object {
            private val pattern = Regex("""(mask|mem)(?:\[(\d+)])? = ([X\d]+)""")
            fun fromString(s: String): Instruction {
                val (_, command, register, value) = pattern
                    .find(s)
                    ?.groupValues
                    ?: throw Exception("regex pattern not found in string")
                return when (command) {
                    "mask" -> {
                        val makeMask = { predicate: (Char) -> Boolean ->
                            value.foldRightIndexed(0L) { index, c, acc ->
                                acc + if (predicate(c)) {
                                    1L.shl(value.length - index - 1)
                                } else 0L
                            }
                        }
                        val oneMask = makeMask { it == '1' }
                        val zeroMask = makeMask { it != '0' }
                        val xMask = makeMask { it == 'X' }
                        Mask(oneMask, zeroMask, xMask)
                    }
                    else -> {
                        Mem(register.toLong(), value.toLong())
                    }
                }
            }
        }

        data class Mask internal constructor(val oneMask: Long, val zeroMask: Long, val xMask: Long) : Instruction()

        data class Mem(val register: Long, val value: Long) : Instruction() {
            fun maskedRegisters(oneMask: Long, xMask: Long): List<Long> {
                val oneApplied = register.or(oneMask)
                return (0..35).fold(listOf(0L)) { acc, place ->
                    when {
                        1L.and(xMask.shr(place)) == 1L -> {
                            acc.flatMap { listOf(it, it + 1L.shl(place)) }
                        }
                        1L.and(oneApplied.shr(place)) == 1L -> {
                            acc.map { it + 1L.shl(place) }
                        }
                        else -> {
                            acc
                        }
                    }
                }
            }

            fun maskedValue(oneMask: Long, zeroMask: Long): Long {
                return value.or(oneMask).and(zeroMask)
            }
        }
    }

    private val instructions = input
        .lines()
        .map(Instruction::fromString)

    override fun part1(): Long {
        val registers = mutableMapOf<Long, Long>()
        var oneMask = 0L
        var zeroMask = 0L
        instructions.forEach { instruction ->
            when (instruction) {
                is Instruction.Mask -> {
                    oneMask = instruction.oneMask
                    zeroMask = instruction.zeroMask
                }

                is Instruction.Mem -> {
                    registers[instruction.register] = instruction.maskedValue(oneMask, zeroMask)
                }
            }
        }
        return registers.values.sum()
    }

    override fun part2(): Long {
        val registers = mutableMapOf<Long, Long>()
        var oneMask = 0L
        var xMask = 0L
        instructions.forEach { instruction ->
            when (instruction) {
                is Instruction.Mask -> {
                    oneMask = instruction.oneMask
                    xMask = instruction.xMask
                }

                is Instruction.Mem -> {
                    instruction.maskedRegisters(oneMask, xMask).forEach {
                        registers[it] = instruction.value
                    }
                }
            }
        }
        return registers.values.sum()
    }
}

fun main() = Day.benchmarkDay(Y20D14::class)

//    Class creation: 34ms
//    Part 1: 11926135976176 (0ms)
//    Part 2: 4330547254348 (83ms)
//    Total time: 119ms

File: rust/advent/src/bin/2020/y20d14.rs
use std::collections::HashMap;

use advent::utilities::get_input::get_input;
use itertools::Itertools;
use utilities::{
    parsing::get_numbers::ContainsNumbers,
    structs::stopwatch::{ReportDuration, Stopwatch},
};

type Input = Vec<Instruction>;
type Output = u64;

fn main() {
    let mut stopwatch = Stopwatch::new();
    stopwatch.start();
    let input = get_input(20, 14).unwrap();
    let input = parse_input(&input);
    println!("Input parsed ({})", stopwatch.lap().report());
    println!("1. {} ({})", part1(&input), stopwatch.lap().report());
    println!("2. {} ({})", part2(&input), stopwatch.lap().report());
    println!("Total: {}", stopwatch.stop().report());
}

enum Instruction {
    Mask { one: u64, zero: u64, x: u64 },
    Mem { register: u64, value: u64 },
}

impl From<&str> for Instruction {
    fn from(value: &str) -> Self {
        let (id, remaining) = value.split_once(['[', ' ']).unwrap();
        match id {
            "mask" => {
                let (_, value) = remaining.split_at(2);
                let make_mask = |predicate: fn(u8) -> bool| {
                    value
                        .as_bytes()
                        .iter()
                        .enumerate()
                        .rfold(0, |acc, (idx, &b)| {
                            if predicate(b) {
                                acc + (1 << (value.len() - idx - 1))
                            } else {
                                acc
                            }
                        })
                };
                let one = make_mask(|b| b == b'1');
                let zero = make_mask(|b| b != b'0');
                let x = make_mask(|b| b == b'X');
                Self::Mask { one, zero, x }
            }
            "mem" => {
                let (register, value) = remaining.get_numbers().collect_tuple().unwrap();
                Self::Mem { register, value }
            }
            s => {
                panic!("{s} not a recognized instruction type");
            }
        }
    }
}

fn masked_registers(register: u64, one_mask: u64, x_mask: u64) -> Vec<u64> {
    let one_applied = register | one_mask;
    (0..36).fold(vec![0], |acc, place| {
        if 1 & (x_mask >> place) == 1 {
            acc.into_iter().flat_map(|n| [n, n + (1 << place)]).collect()
        } else if 1 & (one_applied >> place) == 1 {
            acc.into_iter().map(|n| n + (1 << place)).collect()
        } else {
            acc
        }
    })
}

fn masked_value(value: u64, one_mask: u64, zero_mask: u64) -> u64 {
    (value | one_mask) & zero_mask
}

fn parse_input(input: &str) -> Input {
    input.lines().map(Instruction::from).collect()
}

fn part1(instructions: &Input) -> Output {
    let mut registers = HashMap::new();
    let mut one_mask = 0;
    let mut zero_mask = 0;
    for instruction in instructions {
        match *instruction {
            Instruction::Mask { one, zero, x: _ } => {
                one_mask = one;
                zero_mask = zero;
            }
            Instruction::Mem { register, value } => {
                registers.insert(register, masked_value(value, one_mask, zero_mask));
            }
        }
    }
    registers.values().copied().sum()
}

fn part2(instructions: &Input) -> Output {
    let mut registers = HashMap::new();
    let mut one_mask = 0;
    let mut x_mask = 0;
    for instruction in instructions {
        match *instruction {
            Instruction::Mask { one, zero: _, x } => {
                one_mask = one;
                x_mask = x;
            }
            Instruction::Mem { register, value } => {
                for reg in masked_registers(register, one_mask, x_mask) {
                    registers.insert(reg, value);
                }
            }
        }
    }
    registers.values().copied().sum()
}

#[test]
fn default() {
    let input = get_input(20, 14).unwrap();
    let input = parse_input(&input);
    assert_eq!(11926135976176, part1(&input));
    assert_eq!(4330547254348, part2(&input));
}

// Input parsed (75μs)
// 1. 11926135976176 (38μs)
// 2. 4330547254348 (4ms)
// Total: 4ms